src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -3056,7 +3091,280 @@ public static void Main()
+  IL_0027:  ldfld      ""int System.ValueTuple<int, int>.Item2""
+  IL_002c:  ldc.i4.0
+  IL_002d:  ceq
+  IL_002f:  br.s       IL_0032
+  IL_0031:  ldc.i4.0
+  IL_0032:  call       ""void System.Console.Write(bool)""
+  IL_0037:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TestEqualOnNullableVsNullableTuples_Tuple_NeverNull_AlwaysNull()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        System.Console.Write((1, 2) == ((int, int)?)null);
+    }
+}
+";
+
+            CompileAndVerify(source, expectedOutput: "False", options: TestOptions.ReleaseExe).VerifyIL("C.Main", @"{
+  // Code size        7 (0x7)
+  .maxstack  2
+  IL_0000:  ldc.i4.0
+  IL_0001:  call       ""void System.Console.Write(bool)""
+  IL_0006:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TestEqualOnNullableVsNullableTuples_Tuple_AlwaysNull_MaybeNull()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        M(null);
+    }
+    public static void M((int, int)? t)
+    {
+        System.Console.Write(((int, int)?)null == t);
+    }
+}
+";
+
+            CompileAndVerify(source, expectedOutput: "True", options: TestOptions.ReleaseExe).VerifyIL("C.M", @"{
+  // Code size       21 (0x15)
+  .maxstack  2
+  .locals init ((int, int)? V_0,
+                System.ValueTuple<int, int> V_1)
+  IL_0000:  ldarg.0
+  IL_0001:  stloc.0
+  IL_0002:  ldloca.s   V_0
+  IL_0004:  call       ""bool (int, int)?.HasValue.get""
+  IL_0009:  brfalse.s  IL_000e
+  IL_000b:  ldc.i4.0
+  IL_000c:  br.s       IL_000f
+  IL_000e:  ldc.i4.1
+  IL_000f:  call       ""void System.Console.Write(bool)""
+  IL_0014:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TestEqualOnNullableVsNullableTuples_Tuple_AlwaysNull_NeverNull()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        System.Console.Write(((int, int)?)null == (1, 2));
+    }
+}
+";
+
+            CompileAndVerify(source, expectedOutput: "False", options: TestOptions.ReleaseExe).VerifyIL("C.Main", @"{
+  // Code size        7 (0x7)
+  .maxstack  2
+  IL_0000:  ldc.i4.0
+  IL_0001:  call       ""void System.Console.Write(bool)""
+  IL_0006:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TestEqualOnNullableVsNullableTuples_ElementAlwaysNull()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        System.Console.Write((null, null) == (new int?(), new int?()));
+    }
+}
+";
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "True");
+            verifier.VerifyIL("C.Main", @"{
+  // Code size        9 (0x9)
+  .maxstack  1
+  IL_0000:  nop
+  IL_0001:  ldc.i4.1
+  IL_0002:  call       ""void System.Console.Write(bool)""
+  IL_0007:  nop
+  IL_0008:  ret
+}
+");
```

Comment:
:memo: The never-null vs. never-null is covered by `TestEqualOnNullableVsNullableTuples_NeverNull`.
The maybe-null vs. maybe-null is covered by many tests in this file.